spec/models/concerns/operations/order_fulfilment_operations_spec.rb

```diff
@@ -0,0 +1,89 @@
+require 'rails_helper'
+
+context OrderFulfilmentOperations do
+
+  describe 'Undispatching an orders_package' do
+    let(:dispatch_location) { create(:location, :dispatched) }
+    let(:pkg) { create(:package, received_quantity: 30) }
+    let(:location) { create(:location) }
+    let!(:dispatch_pkg_loc) { create(:packages_location, package: pkg, location: dispatch_location, quantity: 30) }
+    let!(:orders_package) { create(:orders_package, :with_state_dispatched, package: pkg, quantity: 30) }
+    let(:subject) {
+      Class.new { include OrderFulfilmentOperations }
+    }
+
+    before { subject::Operations::undispatch(orders_package, to_location: location) }
+
+    it 'removes the dispatched packages_location record' do
+      expect(PackagesLocation.find_by(id: dispatch_pkg_loc.id)).to be_nil
+    end
+
+    it 'adds the quantity from the dest location' do
+      expect(
+        PackagesLocation.find_by(location: location, package: pkg).quantity
+      ).to eq(30)
+    end
+
+    it 'sets the state to designated' do
+      expect(orders_package.reload.state).to eq('designated')
+    end
+  end
+
+  describe 'Dispatching an orders_package' do
+    let!(:dispatch_location) { create(:location, :dispatched) }
+    let(:pkg) { create(:package, received_quantity: 30) }
+    let(:location) { create(:location) }
+    let!(:pkg_loc) { create(:packages_location, package: pkg, location: location, quantity: 30) }
+
+    [
+      :submitted,
+      :draft,
+      :processing,
+      :closed,
+      :cancelled
```

Comment:
Can you use `Order::ORDER_UNPROCESSED_STATES`?

Reply:
Thought about it, but I'm trying to force myself to separate specs from implementation.
e.g Say someone mistakenly adds a bad state to the constant, we still need the test to fail :)